### PR TITLE
feat(sdk): Improve dashboard to question navigation

### PR DESCRIPTION
**Pull request by deniskaber**
Closes https://github.com/metabase/metabase/issues/44153

### Description

This adds a bit better version of a Question header after navigating from a dashboard.
Also this adds dashboard state preservation after navigating back from a question.

### Demo

https://github.com/metabase/metabase/assets/7983744/a51f1bf6-0f34-4d4b-a9ec-f30ce4b2d506

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/view/ViewHeader/components/DashboardBackButton/DashboardBackButton.tsx [hunk 1]
```diff
@@ -29,8 +39,12 @@ export function DashboardBackButton() {
     <Tooltip tooltip={label}>
       <BackButtonContainer>
         <BackButton
-          as={Link}
-          to={Urls.dashboard(dashboard)}
+          {...(noLink
```
Comment (heypoom) [inline frontend/src/metabase/query_builder/components/view/ViewHeader/components/DashboardBackButton/DashboardBackButton.tsx:42]: Is there a scenario where we would want to apply onClick without also applying noLink? It seems to me like we would always require `noLink` to be `true` anytime an onClick handler is provided, so the condition here could also be `onClick ? {} : ...`.

I think the current code is fine if we want to be explicit about disabling links though.

File: frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx [hunk 2]
```diff
@@ -98,20 +99,22 @@ class PublicOrEmbeddedDashboardInner extends Component<PublicOrEmbeddedDashboard
       dashboardId,
     } = this.props;
 
-    initialize();
+    if (!dashboard) {
+      initialize();
```
Comment (heypoom) [inline frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx:103]: nice, I think this prevents us from initializing twice. I wonder if we should have a test case for this, just to prevent regressions?

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestionResult.tsx [hunk 3]
```diff
@@ -91,16 +103,26 @@ export const InteractiveQuestionResult = ({
...
             (customTitle || (
               <h2 className={cx(CS.h2, CS.textWrap)}>
-                {question.displayName()}
+                {isSaved ? (
+                  question.displayName()
+                ) : isNative ? (
+                  t`New question`
+                ) : (
+                  <QuestionDescription question={question} />
+                )}
```
Comment (heypoom) [inline enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestionResult.tsx:120]: nit: if you want to avoid nested ternaries, you can also either move this into a `useMemo` or use `ts-pattern`.

in this case it's a single-variable condition, so it's already quite clear though.

**Review by heypoom — APPROVED** [review]
The UX looks great!

**Comment by WiNloSt** [comment]
This comment is not directly related to the change in this PR, but are we planning to fix the loading position when drilling? I find it jarring when looking at the demo since my eye are mostly in the middle of the dashboard then, then the loading spinner appear in the top left of the dashboard container.

**Review by WiNloSt — CHANGES_REQUESTED** [review]
Let me test a few things I think the dashboard doesn't load another dashboard after loading one dashboard.

File: frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx [hunk 4]
```diff
@@ -98,20 +99,22 @@ class PublicOrEmbeddedDashboardInner extends Component<PublicOrEmbeddedDashboard
       dashboardId,
     } = this.props;
 
-    initialize();
+    if (!dashboard) {
```
Comment (WiNloSt) [inline frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx:102]: I think this block has a problem, I'm trying to trace the code, but from my testing, if you go to overview page in Shoppy and select other dashboard, only the first dashboard that was loaded will be shown
image

Yeah, I've tested this 2 times. I'm pretty sure this causes the problem. The reason is that `dashboard` is from redux, when we want to fetch a new dashboard the only thing that's change is `dashboardId` so this is where we call `_initialize` to fetch a new dashboard again
https://github.com/metabase/metabase/blob/bf7dc9829ec33d77afbeec89754e32d6da83f570/frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx#L134-L137

if we have this if block, it will never fetch a new dashboard because `dashboard` here is already present, from a redux store, and the result is the old dashboard will be shown. You need to remove this if block.

File: frontend/src/metabase/query_builder/components/view/ViewHeader/ViewTitleHeader.tsx [hunk 5]
```diff
@@ -134,7 +135,6 @@ export function ViewTitleHeader({
     }
   }, [previousQuestion, question, expandFilters, previousQuery, query]);
 
-  const { isNative } = Lib.queryDisplayInfo(query);
```
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/query_builder/components/view/ViewHeader/ViewTitleHeader.tsx]: I see that this change is reverted and we keep this line here.

File: frontend/src/metabase/query_builder/components/view/ViewHeader/components/DashboardBackButton/DashboardBackButton.tsx [hunk 6]
```diff
@@ -11,12 +11,24 @@ import {
 } from "metabase/query_builder/components/view/ViewHeader/ViewTitleHeader.styled";
 import { getDashboard } from "metabase/query_builder/selectors";
 
-export function DashboardBackButton() {
+type DashboardBackButtonProps =
+  | { noLink?: false; onClick?: () => void }
+  | {
+      noLink: true;
+      onClick: () => void;
+    };
```
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/query_builder/components/view/ViewHeader/components/DashboardBackButton/DashboardBackButton.tsx]: Good point.

File: frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx [hunk 7]
```diff
@@ -98,16 +98,18 @@ class PublicOrEmbeddedDashboardInner extends Component<PublicOrEmbeddedDashboard
       dashboardId,
     } = this.props;
 
-    initialize();
+    if (!this.props.dashboard || isForceUpdate) {
```
Comment (WiNloSt) [inline frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx:101]: I'd rather we use the current approach for the same function on our dashboard app.

For example
https://github.com/metabase/metabase/blob/ee7c434781cf61bce03edc5494700e75283fc092/frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx#L263

This could be achieve here with the same flag, but without wrapping this whole thing inside another if block
```ts
    const clearCache = Boolean(!dashboard || isForceUpdate);

    initialize({ clearCache });

    const result = await fetchDashboard({
      dashId: String(dashboardId),
      queryParams: parameterQueryParams,
      options: {
        clearCache,
      },
    });
    ...
      if (this.props.dashboard?.tabs?.length === 0) {
        await fetchDashboardCardData({ reload: false, clearCache });
      }
```

The goal is to not try to deviate from the counterpart entity we have in the main app, so it's easier to compare feature parity and maintain them.

Comment (WiNloSt) [reply frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx:101]: Apart from the comment above this code still has a problem that it won't load a new dashboard if you unmount and remount the dashboard again. Since `_initialize()` will be called without `true`
https://github.com/metabase/metabase/blob/ee7c434781cf61bce03edc5494700e75283fc092/frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx#L125-L127

I think the logic should be
- By default reload the dashboard
- Only _not_ reloading the dashboard _if_ navigating via back button

We're doing the second one in the main app with this flag
https://github.com/metabase/metabase/blob/ee7c434781cf61bce03edc5494700e75283fc092/frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx#L263

However, we need to achieve the same result here somehow. But using `dashboard` isn't working because we don't know if a render is from navigating back via the back button, or if this is the first render e.g. from choosing the dashboard in the analytics overview tab
image


In short, these 2 comments are 2 separate issues.
1. I want consistent code
2. The code for caching dashboard when navigating back to the same dashboard doesn't quite work.

File: enterprise/frontend/src/embedding-sdk/components/public/InteractiveDashboard/InteractiveDashboard.tsx [hunk 8]
```diff
@@ -48,14 +51,18 @@ const InteractiveDashboardInner = ({
...
-
+  const dispatch = useDispatch();
   const { theme } = useEmbedTheme();
 
   const { font } = useEmbedFont();
 
   const store = useStore();
   const [adhocQuestionUrl, setAdhocQuestionUrl] = useState<string | null>(null);
 
+  useUnmount(() => {
```
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/components/public/InteractiveDashboard/InteractiveDashboard.tsx:62]: There is one case, we missed is when we drilled a dashboard and then navigated from one dashboard to another dashboard without unmounting e.g. When on a dashboard, click a link that will render another dashboard.
Screenshot 2024-06-27 at 12 37 56 AM

We need to clear `adhocQuestionUrl` otherwise, we'll keep showing `InteractiveAdHocQuestion` instead of a new dashboard.

I've implemented that with another look
```ts
  useEffect(() => {
    if (dashboardId) {
      dispatch(reset()); // reset "isNavigatingBackToDashboard" state
      setAdhocQuestionUrl(null);
    }
  }, [dashboardId, dispatch]);
```

I also reset `isNavigatingBackToDashboard` here because from a similar scenario.
- Drill a dashboard
- Navigate back (now `isNavigatingBackToDashboard` = `true`)
- Navigate to a different dashboard without unmounting

The `isNavigatingBackToDashboard` would still be `true` if we don't reset it in the hook above.

That was just a solution I came up with. Please feel free to use any solution that suits your situation more.

File: frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx [hunk 9]
```diff
@@ -93,13 +95,19 @@ class PublicOrEmbeddedDashboardInner extends Component<PublicOrEmbeddedDashboard
       setErrorPage,
       parameterQueryParams,
       dashboardId,
+      isNavigatingBackToDashboard,
     } = this.props;
 
-    initialize();
+    const shouldReloadDashboardData = !isNavigatingBackToDashboard;
+
+    initialize({ clearCache: shouldReloadDashboardData });
```
Comment (WiNloSt) [inline frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx:103]: I find we still need to pass `forceUpdate` like you did in the previous version. Take this scenario.
- Drill a dashboard
- Navigate back to the dashboard
- Navigate to another dashboard without unmounting

Here `isNavigatingBackToDashboard` will always be `true` whether or not we dispatch `reset` anywhere. I've tried dispatching `reset` inside the hook in the comment above. And before we call `_initialize` in the `componentDidUpdate` life cycle. Inside `_initialize` you would still get `isNavigatingBackToDashboard: true` even though `reset` was already dispatched.

I suspect this happens because somehow Redux updates the state asynchronously. I don't know why, I thought by default it should be synchronous, since `reset` is a sync action. Maybe that has something to do with our redux setup, but I wouldn't go that far.

This cause a problem when i'm on a dashboard with an empty tab.
- drill 
- navigate back
- click on an empty tab
- navigate to another dashboard without unmounting

My dashboard won't be loaded.

**Review by WiNloSt — APPROVED** [review]
I left 2 comments with 2 separate but similar scenarios.

Feel free to work on those in a separate PR, if that's too complicated. I haven't took any more time to test more scenarios, but maybe there are more edge cases we haven't thought about. But I think this PR is in a better shape and I feel okay if we were to merge this as is, as we can continue to iterate on those edge cases later on.
